clients/cloud/ruby/producer.rb

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/ruby
+#
+# Copyright 2018 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# =============================================================================
+#
+# Produce messages to Confluent Cloud
+# Using the ZenDesk Ruby Client for Apache Kafka (https://github.com/zendesk/ruby-kafka)
+#
+# =============================================================================
+
+require 'json'
+require './c_cloud'
+
+ccloud = CCloud.new
+topic = ccloud.topic
+begin
+  ccloud.kafka.create_topic(topic, num_partitions: 1, replication_factor: 3)
+  puts "Created topic #{topic}"
+rescue Kafka::TopicAlreadyExists => e
+  puts "Did not create topic #{topic} because it already exists."
+end
+
+produced_messages = 0
+begin
+  0.upto(9).each do |n|
+    record_key = 'alice'
+    record_value = JSON.dump(count: n)
+    record = "#{record_key}\t#{record_value}"
+    puts "Producing record: #{record}"
+
+    begin
+      ccloud.producer.produce(record_value, key: record_key, topic: topic)
+      produced_messages += 1
+    rescue Kafka::Error => e
+      puts "Failed to produce record #{record}: #{e.message}"
+    end
+  end
+ensure
+  # deliver all the buffered messages
+  ccloud.producer.deliver_messages
```

Comment:
how are delivery errors reported?

Reply:
Seems like they're just dropped. The `.shutdown` method actually ensures delivery. Digging into it, it just waits for a Worker thread to shut down when it gives it a `:shutdown` message. Here is how it is handled:
```ruby
when :shutdown
            begin
              # Deliver any pending messages first.
              @producer.deliver_messages
            rescue Error => e
              @logger.error("Failed to deliver messages during shutdown: #{e.message}")

              @instrumenter.instrument("drop_messages.async_producer", {
                message_count: @producer.buffer_size + @queue.size,
              })
            end

            # Stop the run loop.
            break
```

Reply:
Nope
```
If you use the asynchronous producer there's no guarantee that messages will have been delivered after #deliver_messages returns. A way of blocking until a message has been delivered with the asynchronous producer may be implemented in the future.
```

Reply:
That's terrible.